### ISSUE TITLE
mtr shows 100% loss on the end host when port 53 is used for UDP probes

### ISSUE THREAD

**Issue by garrymar** [traviscross/mtr#332]
Dear developers,

Mtr does not handle packets returned by a DNS server when it is used with combination of UDP + port 53.

$ mtr -t -n --udp --port 53 -c 1 -r 10.196.76.1
Start: 2020-01-23T12:53:22+0500
HOST: gentoo.local            Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 192.168.5.1                0.0%     1    1.2   1.2   1.2   1.2   0.0
  2.|-- 192.168.254.5              0.0%     1    2.0   2.0   2.0   2.0   0.0
  3.|-- 172.16.237.13              0.0%     1    2.6   2.6   2.6   2.6   0.0
  4.|-- ???                       100.0     1    0.0   0.0   0.0   0.0   0.0

The troubleshooting showed that a DNS server sends a DNS reply with Reply-code: format error to the port that mtr previously used to send the request, but mtr does not handle this reply, and OS sends back an ICMP Port unreachable to the DNS server.

Thanks.

**Comment by rewolff** [comment]
What you're saying is that MTR uses portA to query a DNS server, and portB to send probes. 
The DNS server then gets the request from portB but then replies to portA with the expected "format error". Then the DNS part of MTR has already shut down the portA and the OS rejects the packet to portA. 

How is the DNS server replying to the wrong port mtr's problem?

**Comment by garrymar** [comment]
No, it's about the probe itself. When MTR sends a UDP probe to port 53 (with all zeroes in data field) of a DNS server, the latter replies with "format error". MTR ignores this reply to its probe.

I don't mean DNS queries used to resolve hops in the output (in fact I always use -n flag).

**Comment by rewolff** [comment]
Oh. Ok. So.. mtr expects to see an "port unreachable" and doesn't pick up on the format error. 

MTR doesn't know all protocols, so when you trace to a DNS server on port 53, that the packet is considered an error is expected. But ANY reply from the destination host should be interpreted as having reached the destination.

**Comment by garrymar** [comment]
> Oh. Ok. So.. mtr expects to see an "port unreachable" and doesn't pick up on the format error.

Exactly, but with a tiny correction. MTR expects to see a "port unreachable" and doesn't take into account a reply from an open UDP port.


> MTR doesn't know all protocols, so when you trace to a DNS server on port 53, that the packet is considered an error is expected.

Of cause, it shouldn't. Any reply from an open UDP service is enough to consider the host to be up.


>But ANY reply from the destination host should be interpreted as having reached the destination.

Exactly!

Thanks.

**Comment by rewolff** [comment]
On the one hand, sending "random data" to a working server" may cause "funny" effects. So you'd normally want to test ports that you don't expect to be open. On the other hand, something in your path might only allow packets through to say... the DNS port of the nameserver... 

Some people have a "strict" setup that have a firewall blocking(dropping!) everything that is not directed to the single port that the server is supposed to be serving. So in that case, mtr returning direct "host up" would be good. 

Anyway, do you see the "format error" packet reaching MTR and mtr ignores it or is the kernel discarding it because mtr faked the outgoing UDP packet and didn't open the actual UDP port it is using? I'm thinking the latter, right?

**Comment by garrymar** [comment]
> Anyway, do you see the "format error" packet reaching MTR and mtr ignores it or is the kernel discarding it because mtr faked the outgoing UDP packet and didn't open the actual UDP port it is using? I'm thinking the latter, right?

To he honest, I don't know how MTR interacts with TCP/IP stack of Linux OS, whether it opens a socket actually or just spoofs source port. If it spoofs source port, then of course, an OS will reject a reply.

**Comment by nomis** [comment]
This makes it very difficult to test a latency issue only affecting DNS replies using mtr.

The OS on the host running mtr is returning port unreachable to all the responses.
